### PR TITLE
Issue #79, #76 & #65. Make HDFS optional. Check HDFS status. Write to local FS

### PR DESCRIPTION
**Pull request by tushargupta98**
Issue #76 Make other_examples work with or without HDFS. binarize_fsl partially fixed. Clarification needed.
Issue #65 Which tmpfs should be used.

### CODE REVIEW

File: other-examples/common_functions.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        # import subprocess
+        print('Running system command: {0}'.format(' '.join(args_list)))
+        proc = subprocess.Popen(args_list, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        s_output, s_err = proc.communicate()
+        s_return =  proc.returncode
+        return s_return, s_output, s_err
+
+
+def hdfs_status():
+    (ret, out, err) = run_cmd(['jps'])
```
Comment (glatard) [inline other-examples/common_functions.py:13]: return False if ret is not 0

File: other-examples/common_functions.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    if ('SecondaryNameNode' not in out) & ('DataNode' not in out) & ('NameNode' not in out):
+        return False
+    else:
+        return True
+# list in hdfs to test if it is working or not
+def hdfs_start():
```
Comment (glatard) [inline other-examples/common_functions.py]: return True or False depending whether hdfs start was successful

File: other-examples/common_functions.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    else:
+        print "HDFS could not be started. Please start it manually."
+
+def directory_exists(dir_path):
+    if os.path.isdir(dir_path) & (os.path.exists(dir_path)):
+        return True
+    else:
+        return False
+
+def tmpfs_directories():
```
Comment (glatard) [inline other-examples/common_functions.py]: take the first one where you can write and delete a file.

File: other-examples/binarize_fsl.py [hunk 4]
```diff
@@ -9,25 +9,39 @@
 import os
 import subprocess
 from hdfs import Config
+
 import sys
+from  common_functions import directory_exists,tmpfs_directories, hdfs_status
 
-def copy_to_hdfs(x, output, client):
+def copy_to_hdfs(x, output, hdfs, tmpfs_path, client):
```
Comment (glatard) [inline other-examples/binarize_fsl.py]: remove this method

Comment (glatard) [reply other-examples/binarize_fsl.py]: and the call to this method in the pipeline.

File: other-examples/binarize_fsl.py [hunk 5]
```diff
@@ -9,25 +9,39 @@
...
+    nib.save(im, os.path.join(tmpfs_path, filename))
     return filename
 
-def binarize(x, threshold):
-    filename = save_nifti(x)
-    run_command = "sudo docker run -v /run/user/1000:/binarized_tmp --rm fsl-docker-img bin/fsl_bin.sh {0} {1}".format(filename, threshold)
 
-    subprocess.check_output(run_command, shell=True, stderr=subprocess.STDOUT)
+def binarize(x, threshold, tmpfs_path):
+    if not(directory_exists(tmpfs_path)):
```
Comment (glatard) [inline other-examples/binarize_fsl.py]: tmpfs_path should be optional. If it is None, you should figure it out using tmpfs_directories.

File: other-examples/binarize_fsl.py [hunk 6]
```diff
@@ -9,25 +9,39 @@
...
+        tmpfs_path = tmpfs_directories()
+    if tmpfs_path != "":
+        filename = save_nifti(x, tmpfs_path)
+        run_command = "sudo docker run -v {2}:/binarized_tmp --rm fsl-docker-img bin/fsl_bin.sh {0} {1}".format(filename, threshold, tmpfs_path)
+        result = subprocess.check_output(run_command, shell=True, stderr=subprocess.STDOUT)
+    else:
```
Comment (glatard) [inline other-examples/binarize_fsl.py]: fall back on current working directory.

File: other-examples/binarize_fsl.py [hunk 7]
```diff
@@ -9,25 +9,39 @@
...
-    filename = save_nifti(x)
-    run_command = "sudo docker run -v /run/user/1000:/binarized_tmp --rm fsl-docker-img bin/fsl_bin.sh {0} {1}".format(filename, threshold)
 
-    subprocess.check_output(run_command, shell=True, stderr=subprocess.STDOUT)
+def binarize(x, threshold, tmpfs_path):
+    if not(directory_exists(tmpfs_path)):
+        tmpfs_path = tmpfs_directories()
+    if tmpfs_path != "":
+        filename = save_nifti(x, tmpfs_path)
+        run_command = "sudo docker run -v {2}:/binarized_tmp --rm fsl-docker-img bin/fsl_bin.sh {0} {1}".format(filename, threshold, tmpfs_path)
```
Comment (glatard) [inline other-examples/binarize_fsl.py]: Fix #80 first

File: other-examples/binarize_spark.py [hunk 8]
```diff
@@ -8,63 +8,75 @@
...
-
     parser = argparse.ArgumentParser(description='Binarize images')
     parser.add_argument('threshold', type=int, help="binarization threshold")
     parser.add_argument('folder_path', type=str, help='folder path containing all of the splits')
     parser.add_argument('output_path', type=str, help='output folder path')
-    parser.add_argument('num', type=int,choices=[2,4,6,8], help='number of binarization operations')
-    parser.add_argument('-m', '--in_memory', type=bool, default=True,  help='in memory computation')    
+    parser.add_argument('num', type=int, choices=[2, 4, 6, 8], help='number of binarization operations')
+    parser.add_argument("--hdfs", action='store_true', help="Use HDFS")
+    parser.add_argument('-m', '--in_memory', type=bool, default=True, help='in memory computation')
```
Comment (glatard) [inline other-examples/binarize_spark.py]: rename this to --tmpfs and only attempt to write to tmpfs when this is on.

**Review by glatard — CHANGES_REQUESTED** [review]

**Comment by tushargupta98** [comment]
Hi,
I have committed the changes, could you please review the same?
Thanks!

File: other-examples/binarize_fsl.py (hunk 7 above)
Comment (tushargupta98) [reply other-examples/binarize_fsl.py]: Except this one, I have addressed the rest of the changes with the latest commit. 
I will include changes for this one with the next pull request. Could you please review the latest changes?
Thanks!

**Comment by tushargupta98** [comment]
jps alias was not found on Travis-ci

**Comment by ValHayot** [comment]
there's no HDFS yet on the docker image. it's spark-only.

**Pull request merged**